python/cudf/cudf/io/parquet.py

```diff
@@ -1010,9 +1010,13 @@ def __init__(
     ) -> None:
         if isinstance(path, str) and path.startswith("s3://"):
             self.fs_meta = {"is_s3": True, "actual_path": path}
-            self.path = tempfile.TemporaryDirectory().name
+            self.dir_: Optional[
```

Comment:
Yes, types in both branches must be the same. That is how mypy always works. Here's an example (I'm using `sys.argv` rather than a trivial `if True` because mypy is smart enough to ignore unreachable branches).
```
# test.py
import sys
if len(sys.argv) > 2:
    x = 1
else:
    x = "hello"
```
results in 
```
(main) dt08% mypy test.py
test.py:5: error: Incompatible types in assignment (expression has type "str", variable has type "int")  [assignment]
```
Note that in this example I didn't even bother annotating the type, mypy just inferred the types and noticed the incompatibility.

It is possible to forward-declare the type, but it's not something I've seen done much and not anywhere in cudf except for class members. I would be open to discussing that change, but it's out of scope for this PR since only doing it here would be inconsistent with typing everywhere else in cudf (which is already fairly inconsistent at the moment).